entrypoint: Enabling bash -x with DEBUG environment variable

ErwanAliasr1 opened the pull request:
If a container fails to start, it's pretty complicated to understand what went wrong and why.
As per the container thing, it's even more difficult as if a command fails, the container stops making any tracing impossible.
If you can't modify the container, then it's pretty difficult to analyse the issue.

This patch offer to "control" the verbosity of bash with a DEBUG environement variable.

If you need to trace a particular container, just set DEBUG to any value at run time like :
    sudo docker run -d --name ceph-mon -e DEBUG=1 -e NETWORK_AUTO_DETECT=1 daemon mon

Later, if we want various behavior, we could read the value of DEBUG and trigger different options.
This first patch makes things easy by just considering the presence and not the value of DEBUG.

leseb commented on ceph-releases/kraken/ubuntu/16.04/daemon/entrypoint.sh:
```diff
@@ -1,6 +1,37 @@
+  set -x
+
+  # Now let's see if we have some special options
+  case "${DEBUG,,}" in
```
Please use the `to_lowercase` function

leseb commented on ceph-releases/kraken/ubuntu/16.04/daemon/common_functions.sh:
```diff
@@ -154,3 +154,31 @@ function is_integer {
+}
+
+do_exec() {
+  # This function do the same kind of behavior as exec
```
s/do/does/

leseb commented on ceph-releases/kraken/ubuntu/16.04/daemon/common_functions.sh:
```diff
@@ -154,3 +154,31 @@ function is_integer {
+  # but it does not really change the process in place
+  # As per docker requirement, we forward the SIGTERM
+  # to the real program.
+  # That is allowing trapping ERR and handling failures
```
s/is allowing/allows/

leseb commented on ceph-releases/kraken/ubuntu/16.04/daemon/entrypoint.sh:
```diff
@@ -1,6 +1,37 @@
+  set -x
+
+  # Now let's see if we have some special options
+  case "$(to_lower ${DEBUG})" in
```
s/to_lower/to_lowercase/

ErwanAliasr1 replied:
fixed

leseb commented on ceph-releases/kraken/ubuntu/16.04/daemon/entrypoint.sh:
```diff
@@ -1,6 +1,37 @@
+
+      # Let's call traerror if we failed on ERR
+      trap 'traperror $? $LINENO $BASH_LINENO "$BASH_COMMAND" $(printf "::%s" ${FUNCNAME[@]})' ERR
+      echo "STAYALIVE: container will not die if a command fail"
```
s/fail/fails/

ErwanAliasr1 replied:
fixed

ktdreyer commented on ceph-releases/kraken/ubuntu/16.04/daemon/entrypoint.sh:
```diff
@@ -1,6 +1,37 @@
+  echo "An issue occured and you asked me to stay alive."
+  echo "You can connect to me with : sudo docker exec -i -t $HOSTNAME /bin/bash"
+  echo "When debugging is over just suicide me with : pkill sleep"
+  echo "I'll sleep for 365 days waiting at you darling, bye bye"
```
this last line is bizarre

ErwanAliasr1 commented:
I reworked totally that stuff and split it in two part. One if a file that every container can include to get the 'stayalive' feature. The other patch is about implementing DEBUG= handling.

ErwanAliasr1 commented:
Rebased & ready for merge
